File: backend/src/database/Task.ts
export default class Task {
  public id: number;
  public userId: string;
  public categoryId: number | null;
  public name: string;
  public description: string | null;
  public creationDate: string;
  public dueDate: string | null;
  public completionDate: string | null;

  constructor(
    id: number,
    userId: string,
    categoryId: number | null,
    name: string,
    description: string | null,
    creationDate: string,
    dueDate: string | null,
    completionDate: string | null,
  ) {
    this.id = id;
    this.userId = userId;
    this.categoryId = categoryId;
    this.name = name;
    this.description = description;
    this.creationDate = creationDate;
    this.dueDate = dueDate;
    this.completionDate = completionDate;
  }

  /**
   * Create a task object from a database row.
   */
  public static fromRow(row: any): Task {
    return new Task(
      row.id as number,
      row.user_id as string,
      row.category_id as number | null,
      row.name as string,
      row.description as string | null,
      row.creation_date as string,
      row.due_date as string | null,
      row.completion_date as string | null,
    );
  }

  public toString(): string {
    return `Task [id=${this.id}, userId=${this.userId}, categoryId=${this.categoryId}, name=${this.name}, description=${this.description}, creationDate=${this.creationDate}, dueDate=${this.dueDate}, completionDate=${this.completionDate}]`;
  }

  public toJSON(): any {
    return {
      id: this.id,
      user_id: this.userId,
      category_id: this.categoryId,
      name: this.name,
      description: this.description,
      creation_date: this.creationDate,
      due_date: this.dueDate,
      completion_date: this.completionDate,
    };
  }
}

File: frontend/lib/types/task.dart
import 'category.dart';

/// Public class [Task] represents a task created by the user.
class Task {
  /// The id of the task.
  int? id;

  /// The category of the task.
  Category? category;

  /// The name of the task.
  String? name;

  /// Description of the task.
  String? description;

  /// When the task was created.
  DateTime? creationDate;

  /// The due date of the task.
  DateTime? dueDate;

  /// When the task was completed.
  DateTime? completionDate;

  /// Whether the task is completed or not.
  bool get isCompleted => completionDate != null;

  /// Whether the task has a due date or not.
  bool get hasDueDate => dueDate != null;

  /// Constructor for the [Task] class.
  Task({
    this.id,
    this.category,
    this.name,
    this.description,
    this.creationDate,
    this.dueDate,
    this.completionDate,
  });

  /// Constructor for the [Task] class from a JSON object.
  factory Task.fromJson(Map<String, dynamic> json) => Task(
        id: json['id'],
        category: Category.fromJson(json['category_id']),
        name: json['name'],
        description: json['description'],
        creationDate: DateTime.tryParse(json['creation_date']),
        dueDate: json['due_date'] == null
            ? null
            : DateTime.tryParse(json['due_date']),
        completionDate: json['completion_date'] == null
            ? null
            : DateTime.tryParse(json['completion_date']),
      );

  /// Converts the [Task] object to a JSON object.
  Map<String, dynamic> toJson() => {
        'id': id,
        'category': category?.toJson(),
        'name': name,
        'description': description,
        'creation_date': creationDate?.toIso8601String(),
        'due_date': dueDate?.toIso8601String(),
        'completion_date': completionDate?.toIso8601String(),
      };
}
